Automattic/simplenote-macos
# NIB(s) Localization Support

## Description:
In this PR we're extending the Localization toolchain in order to support Xcode's standard NIB localization mechanism.
This translates into:

1. Support for **strings extraction** from `MainMenu.xib`
2. Support for **downloading MainMenu strings**
3. Enhanced failsafe mechanism to deal with missing GlotPress Strings

@jkmassel Sir, can I bug you with a fun review?

/cc @mokagio @loremattei 

Thank you all!

Ref. #8 

## Details:

### localize.py:
This script now also extracts strings from **MainMenu.nib**, and converts them into UTF-8.


### update-translations.rb
We're now **also** downloading strings from the [Main Menu](https://translate.wordpress.com/projects/simplenote/macos/main-menu/) GlotPress subproject.


### fix-translations:
Missing translations, when downloaded from GlotPress, would look like this:

> "Some Text" = "";

The script `fix-translations` works by setting a `Translation.Value = Translation.Key`, so that the app doesn't display a blank value:

> "Some Text" = "Some Text";

#### Problem: 
NIB translation keys don't follow the same principle. Instead, they contain keypaths:
> "58.title" = "Some Title";

The side effect is: missing NIB translations (in GlotPress) look like this:
> "58.title" = "";

Running the standard `fix-translations` script would yield:
> "58.title" = "58.title";

And this is definitely not human readable 😞 

#### Solution:
In this PR we're enhancing `fix-translations`, so that the actual `fallback value` is extracted from a "Fallback Translations Map". That means: the script will extract the proper "Fallback Translation" from another file.


--

## Test: Localize!
1. Checkout commit 763e2de
2. Run `./Scripts/localize.py`

- [x] Verify `Simplenote/en.lproj/Localizable.strings` has an updated string
- [x] Verify `Simplenote/en.lproj/MainMenu.strings` has an updated string

## Test: Update Strings
1. Run `./Scripts/update-translations.rb`

- [x] Verify `Simplenote/pt-BR.lproj/MainMenu.strings` gets new Portuguese strings (an OSS contributor already did this! yay!)
- [x] Verify (lots of) `Localizable.strings` files get updated
- [x] Verify that `MainMenu.strings` files **don't contain empty Localizations.** Strings that aren't in Glotpress should carry the English localization!

## Release
These changes do not require release notes.
